pkg/sync/sync_context.go

```diff
@@ -921,29 +921,38 @@ func (sc *syncContext) applyObject(t *syncTask, dryRun, force, validate bool) (c
 	var err error
 	var message string
 	shouldReplace := sc.replace || resourceutil.HasAnnotationOption(t.targetObj, common.AnnotationSyncOptions, common.SyncOptionReplace)
-	if shouldReplace {
-		if t.liveObj != nil {
-			// Avoid using `kubectl replace` for CRDs since 'replace' might recreate resource and so delete all CRD instances.
-			// The same thing applies for namespaces, which would delete the namespace as well as everything within it,
-			// so we want to avoid using `kubectl replace` in that case as well.
-			if kube.IsCRD(t.targetObj) || t.targetObj.GetKind() == kubeutil.NamespaceKind {
-				update := t.targetObj.DeepCopy()
-				update.SetResourceVersion(t.liveObj.GetResourceVersion())
-				_, err = sc.resourceOps.UpdateResource(context.TODO(), update, dryRunStrategy)
-				if err == nil {
-					message = fmt.Sprintf("%s/%s updated", t.targetObj.GetKind(), t.targetObj.GetName())
-				} else {
-					message = fmt.Sprintf("error when updating: %v", err.Error())
-				}
-			} else {
-				message, err = sc.resourceOps.ReplaceResource(context.TODO(), t.targetObj, dryRunStrategy, force)
+	applyFn := func(dryRunStrategy cmdutil.DryRunStrategy) (string, error) {
+		if !shouldReplace {
+			return sc.resourceOps.ApplyResource(context.TODO(), t.targetObj, dryRunStrategy, force, validate, serverSideApply, sc.serverSideApplyManager)
+		}
+		if t.liveObj == nil {
+			return sc.resourceOps.CreateResource(context.TODO(), t.targetObj, dryRunStrategy, validate)
+		}
+		// Avoid using `kubectl replace` for CRDs since 'replace' might recreate resource and so delete all CRD instances.
+		// The same thing applies for namespaces, which would delete the namespace as well as everything within it,
+		// so we want to avoid using `kubectl replace` in that case as well.
+		if kube.IsCRD(t.targetObj) || t.targetObj.GetKind() == kubeutil.NamespaceKind {
+			update := t.targetObj.DeepCopy()
+			update.SetResourceVersion(t.liveObj.GetResourceVersion())
+			_, err = sc.resourceOps.UpdateResource(context.TODO(), update, dryRunStrategy)
+			if err == nil {
```

Comment:
As a matter of convention, I'd do `if err != nil` and reverse the actions, but I don't think it matters much.